### PR TITLE
Do not unlock exam programming exercises on edit

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
There is a small logic error in the code when editing exercises it can happen that the student repos are unlocked

### Description
Fix the logic issue

### Steps for Testing
1. Create an exam with programming exercises
2. Participate as student
3. Check that the student repository of the programming exercise is locked after the exam (i.e. the student only has read access)
4. Edit the programming exercises and wait shortly
5. Double check that the student repos is still locked for the student (i.e. the student only has read and **not** write access)

### CODE REVIEW

**Review by sleiss — APPROVED** [review]
Code looks good as far as I can tell, have not tested it.

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java [hunk 1]
```diff
@@ -178,14 +178,20 @@ private void scheduleExamExercise(ProgrammingExercise exercise) {
...
         }
-        else if (examDateService.getLatestIndividualExamEndDate(exam).isBefore(ZonedDateTime.now())) {
+        // DURING EXAM
+        else if (examDateService.getLatestIndividualExamEndDate(exam).isAfter(ZonedDateTime.now())) {
             // This is only a backup (e.g. a crash of this node and restart during the exam)
+            // TODO: Christian Femers: this can lead to a weired edge case after the normal exam end date and before the last individual exam end date (in case of working time
+            // extensions)
             scheduleService.scheduleTask(exercise, ExerciseLifecycle.RELEASE, Set.of(new Tuple<>(ZonedDateTime.now().plusSeconds(5), unlockAllStudentRepositories(exercise))));
         }
+        // NOTHING TO DO AFTER EXAM
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java:194]: (this is not entirely correct, see `BUILD_AND_TEST_AFTER_DUE_DATE` below)

**Review by MaisiKoleni — APPROVED** [review]
Code looks good to me.

**Review by tiffca — APPROVED** [review]
Tested successfully on ts3

**Review by airelawaleria — APPROVED** [review]

**Review by EneaGore — APPROVED** [review]
Seemed to work as expected on ts3.

**Pull request merged**